apps/api/src/deployment/services/top-up-managed-deployments/top-up-managed-deployments.service.ts

```diff
@@ -1,59 +1,67 @@
-import { AllowanceHttpService, DeploymentAllowance } from "@akashnetwork/http-sdk";
 import { LoggerService } from "@akashnetwork/logging";
 import { singleton } from "tsyringe";
 
-import { InjectSigningClient } from "@src/billing/providers/signing-client.provider";
+import { BillingConfig, InjectBillingConfig } from "@src/billing/providers";
 import { InjectWallet } from "@src/billing/providers/wallet.provider";
-import { MasterSigningClientService, MasterWalletService } from "@src/billing/services";
+import { UserWalletOutput, UserWalletRepository } from "@src/billing/repositories";
+import { MasterWalletService, RpcMessageService } from "@src/billing/services";
+import { BalancesService } from "@src/billing/services/balances/balances.service";
+import { TxSignerService } from "@src/billing/services/tx-signer/tx-signer.service";
 import { ErrorService } from "@src/core/services/error/error.service";
-import { DrainingDeploymentOutput } from "@src/deployment/repositories/lease/lease.repository";
 import { DrainingDeploymentService } from "@src/deployment/services/draining-deployment/draining-deployment.service";
+import { DeploymentsRefiller, TopUpDeploymentsOptions } from "@src/deployment/types/deployments-refiller";
 
 @singleton()
-export class TopUpManagedDeploymentsService {
+export class TopUpManagedDeploymentsService implements DeploymentsRefiller {
   private readonly CONCURRENCY = 10;
 
   private readonly logger = new LoggerService({ context: TopUpManagedDeploymentsService.name });
 
   constructor(
-    private readonly allowanceHttpService: AllowanceHttpService,
-    @InjectWallet("MANAGED") private readonly managedMasterWalletService: MasterWalletService,
-    @InjectSigningClient("MANAGED") private readonly managedMasterSigningClientService: MasterSigningClientService,
+    private readonly userWalletRepository: UserWalletRepository,
+    private readonly txSignerService: TxSignerService,
+    @InjectBillingConfig() private readonly billingConfig: BillingConfig,
     private readonly drainingDeploymentService: DrainingDeploymentService,
+    @InjectWallet("MANAGED") private readonly managedMasterWalletService: MasterWalletService,
+    private readonly balancesService: BalancesService,
+    private readonly rpcClientService: RpcMessageService,
     private readonly errorService: ErrorService
   ) {}
 
-  async topUpDeployments() {
-    const address = await this.managedMasterWalletService.getFirstAddress();
-    await this.allowanceHttpService.paginateDeploymentGrants({ granter: address, limit: this.CONCURRENCY }, async grants => {
+  async topUpDeployments(options: TopUpDeploymentsOptions) {
+    await this.userWalletRepository.paginate({ limit: this.CONCURRENCY }, async wallets => {
```

Comment:
Ideally, we should iterate over only the wallets that need top up, because otherwise we have to iterate over all and query deployments for each which will be heavy for the database.

Reply:
Just dropping a response discussed in private for visibility.

there's no concurrency for this job, so it's a single query per several seconds. It uses the index on the query. So it's not expensive.
This is an approach we discussed concluding this is efficient enough for the time being.